File: python/citation_tracker.py
"""
Citation Tracker for AI-Generated Content
This module provides utilities to track and properly attribute sources when using
AI-assisted code generation tools like GitHub Copilot.
"""
import json
import os
from typing import Dict, List, Optional, Union
from datetime import datetime


class CitationTracker:
    """A class for tracking citations in AI-generated content."""
    
    def __init__(self, project_path: str = None):
        """
        Initialize the citation tracker.
        
        Args:
            project_path: Root directory of the project for storing citation data
        """
        self.project_path = project_path or os.getcwd()
        self.citations_file = os.path.join(self.project_path, 'citations.json')
        self.citations = self._load_citations()
        
    def _load_citations(self) -> Dict:
        """Load existing citations from the citations file."""
        if os.path.exists(self.citations_file):
            try:
                with open(self.citations_file, 'r', encoding='utf-8') as f:
                    return json.load(f)
            except json.JSONDecodeError:
                print(f"Warning: Citations file corrupted, creating new one.")
        
        # Initialize a new citations structure
        return {
            "project_info": {
                "name": os.path.basename(self.project_path),
                "created_at": datetime.now().isoformat()
            },
            "sources": {},
            "file_citations": {}
        }
    
    def add_source(self, 
                  source_id: str, 
                  name: str, 
                  url: Optional[str] = None, 
                  author: Optional[str] = None,
                  license_type: Optional[str] = None,
                  description: Optional[str] = None) -> None:
        """
        Add a new citation source.
        
        Args:
            source_id: Unique identifier for the source
            name: Name of the source (e.g., library name, article title)
            url: URL to the source
            author: Author of the source
            license_type: License type (e.g., MIT, Apache 2.0)
            description: Brief description of the source
        """
        self.citations["sources"][source_id] = {
            "name": name,
            "url": url,
            "author": author,
            "license_type": license_type,
            "description": description,
            "added_at": datetime.now().isoformat()
        }
        self._save_citations()
        
    def cite_in_file(self, file_path: str, source_ids: Union[str, List[str]], 
                    line_start: Optional[int] = None, 
                    line_end: Optional[int] = None,
                    comment: Optional[str] = None) -> None:
        """
        Record that a file uses content from specified sources.
        
        Args:
            file_path: Path to the file
            source_ids: Source ID or list of source IDs
            line_start: Optional starting line number
            line_end: Optional ending line number
            comment: Optional comment about how the source was used
        """
        rel_path = os.path.relpath(file_path, self.project_path)
        
        if isinstance(source_ids, str):
            source_ids = [source_ids]
            
        if rel_path not in self.citations["file_citations"]:
            self.citations["file_citations"][rel_path] = []
            
        citation = {
            "source_ids": source_ids,
            "cited_at": datetime.now().isoformat()
        }
        
        if line_start is not None:
            citation["line_start"] = line_start
        if line_end is not None:
            citation["line_end"] = line_end
        if comment:
            citation["comment"] = comment
            
        self.citations["file_citations"][rel_path].append(citation)
        self._save_citations()
        
    def generate_attribution_comment(self, source_id: str) -> str:
        """
        Generate a code comment for attribution.
        
        Args:
            source_id: Source ID to generate comment for
            
        Returns:
            Formatted comment string for attribution
        """
        if source_id not in self.citations["sources"]:
            return f"# Attribution: Unknown source ({source_id})"
            
        source = self.citations["sources"][source_id]
        comment = f"# Attribution: {source['name']}"
        
        if source.get("author"):
            comment += f", by {source['author']}"
        if source.get("url"):
            comment += f"\n# Source: {source['url']}"
        if source.get("license_type"):
            comment += f"\n# License: {source['license_type']}"
            
        return comment
        
    def insert_attribution_header(self, file_path: str) -> str:
        """
        Generate an attribution header for a file.
        
        Args:
            file_path: Path to the file
            
        Returns:
            Formatted header with all attributions for the file
        """
        rel_path = os.path.relpath(file_path, self.project_path)
        if rel_path not in self.citations["file_citations"]:
            return "# No recorded attributions for this file."
            
        file_citations = self.citations["file_citations"][rel_path]
        cited_sources = set()
        for citation in file_citations:
            cited_sources.update(citation["source_ids"])
            
        header = [
            "\"\"\"",
            "This file contains code derived from or inspired by the following sources:",
            ""
        ]
        
        for source_id in cited_sources:
            if source_id in self.citations["sources"]:
                source = self.citations["sources"][source_id]
                header.append(f"- {source['name']}")
                if source.get("url"):
                    header[-1] += f" ({source['url']})"
                if source.get("author"):
                    header.append(f"  Author: {source['author']}")
                if source.get("license_type"):
                    header.append(f"  License: {source['license_type']}")
                header.append("")
        
        header.append("\"\"\"")
        return "\n".join(header)
    
    def generate_citations_markdown(self) -> str:
        """
        Generate a Markdown document with all citations.
        
        Returns:
            Markdown formatted citations document
        """
        md = ["# Project Citations", "", "This document lists all external sources used in this project."]
        
        if not self.citations["sources"]:
            md.append("\n*No citations recorded.*")
            return "\n".join(md)
            
        md.append("\n## Sources\n")
        
        for source_id, source in self.citations["sources"].items():
            md.append(f"### {source['name']}")
            if source.get("author"):
                md.append(f"**Author:** {source['author']}")
            if source.get("url"):
                md.append(f"**URL:** [{source['url']}]({source['url']})")
            if source.get("license_type"):
                md.append(f"**License:** {source['license_type']}")
            if source.get("description"):
                md.append(f"\n{source['description']}")
            md.append("")
            
        md.append("## Usage by File\n")
        
        for file_path, citations in self.citations["file_citations"].items():
            md.append(f"### {file_path}")
            for citation in citations:
                source_names = []
                for source_id in citation["source_ids"]:
                    source = self.citations["sources"].get(source_id, {"name": f"Unknown ({source_id})"})
                    source_names.append(source["name"])
                    
                locations = ""
                if "line_start" in citation and "line_end" in citation:
                    locations = f" (lines {citation['line_start']}-{citation['line_end']})"
                elif "line_start" in citation:
                    locations = f" (from line {citation['line_start']})"
                    
                md.append(f"- Uses: {', '.join(source_names)}{locations}")
                if citation.get("comment"):
                    md.append(f"  - Note: {citation['comment']}")
            md.append("")
            
        return "\n".join(md)
        
    def export_citations_markdown(self, output_path: Optional[str] = None) -> str:
        """
        Export citations to a markdown file.
        
        Args:
            output_path: Optional path for the markdown file
            
        Returns:
            Path to the created markdown file
        """
        if not output_path:
            output_path = os.path.join(self.project_path, "CITATIONS.md")
            
        md_content = self.generate_citations_markdown()
        
        with open(output_path, 'w', encoding='utf-8') as f:
            f.write(md_content)
            
        return output_path
    
    def _save_citations(self) -> None:
        """Save citations to the citations file."""
        with open(self.citations_file, 'w', encoding='utf-8') as f:
            json.dump(self.citations, f, indent=2)

File: csharp/CitationTracker/CitationTracker.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace CitationTracking
{
    /// <summary>
    /// A class for tracking citations in AI-generated content for C# projects.
    /// </summary>
    public class CitationTracker
    {
        private string projectPath;
        private string citationsFile;
        private CitationData citations;

        /// <summary>
        /// Initialize the citation tracker.
        /// </summary>
        /// <param name="projectPath">Root directory of the project for storing citation data</param>
        public CitationTracker(string projectPath = null)
        {
            this.projectPath = projectPath ?? Directory.GetCurrentDirectory();
            this.citationsFile = Path.Combine(this.projectPath, "citations.json");
            this.citations = LoadCitations();
        }

        /// <summary>
        /// Load existing citations from the citations file.
        /// </summary>
        /// <returns>Loaded citation data</returns>
        private CitationData LoadCitations()
        {
            if (File.Exists(citationsFile))
            {
                try
                {
                    string jsonString = File.ReadAllText(citationsFile);
                    return JsonSerializer.Deserialize<CitationData>(jsonString) ?? CreateNewCitationData();
                }
                catch (JsonException)
                {
                    Console.WriteLine("Warning: Citations file corrupted, creating new one.");
                }
            }

            return CreateNewCitationData();
        }

        /// <summary>
        /// Create a new citations data structure.
        /// </summary>
        /// <returns>New citation data object</returns>
        private CitationData CreateNewCitationData()
        {
            return new CitationData
            {
                ProjectInfo = new ProjectInfo
                {
                    Name = Path.GetFileName(projectPath),
                    CreatedAt = DateTime.Now
                },
                Sources = new Dictionary<string, SourceInfo>(),
                FileCitations = new Dictionary<string, List<Citation>>()
            };
        }

        /// <summary>
        /// Add a new citation source.
        /// </summary>
        /// <param name="sourceId">Unique identifier for the source</param>
        /// <param name="name">Name of the source (e.g., library name, article title)</param>
        /// <param name="url">URL to the source</param>
        /// <param name="author">Author of the source</param>
        /// <param name="licenseType">License type (e.g., MIT, Apache 2.0)</param>
        /// <param name="description">Brief description of the source</param>
        public void AddSource(string sourceId, string name, string url = null, 
            string author = null, string licenseType = null, string description = null)
        {
            citations.Sources[sourceId] = new SourceInfo
            {
                Name = name,
                Url = url,
                Author = author,
                LicenseType = licenseType,
                Description = description,
                AddedAt = DateTime.Now
            };
            
            SaveCitations();
        }

        /// <summary>
        /// Record that a file uses content from specified sources.
        /// </summary>
        /// <param name="filePath">Path to the file</param>
        /// <param name="sourceIds">Source ID or array of source IDs</param>
        /// <param name="lineStart">Optional starting line number</param>
        /// <param name="lineEnd">Optional ending line number</param>
        /// <param name="comment">Optional comment about how the source was used</param>
        public void CiteInFile(string filePath, string[] sourceIds, 
            int? lineStart = null, int? lineEnd = null, string comment = null)
        {
            string relPath = Path.GetRelativePath(projectPath, filePath);

            if (!citations.FileCitations.ContainsKey(relPath))
            {
                citations.FileCitations[relPath] = new List<Citation>();
            }

            var citation = new Citation
            {
                SourceIds = sourceIds,
                CitedAt = DateTime.Now,
                LineStart = lineStart,
                LineEnd = lineEnd,
                Comment = comment
            };

            citations.FileCitations[relPath].Add(citation);
            SaveCitations();
        }

        /// <summary>
        /// Record that a file uses content from a specified source.
        /// </summary>
        /// <param name="filePath">Path to the file</param>
        /// <param name="sourceId">Source ID</param>
        /// <param name="lineStart">Optional starting line number</param>
        /// <param name="lineEnd">Optional ending line number</param>
        /// <param name="comment">Optional comment about how the source was used</param>
        public void CiteInFile(string filePath, string sourceId,
            int? lineStart = null, int? lineEnd = null, string comment = null)
        {
            CiteInFile(filePath, new[] { sourceId }, lineStart, lineEnd, comment);
        }

        /// <summary>
        /// Generate a code comment for attribution.
        /// </summary>
        /// <param name="sourceId">Source ID to generate comment for</param>
        /// <returns>Formatted comment string for attribution</returns>
        public string GenerateAttributionComment(string sourceId)
        {
            if (!citations.Sources.ContainsKey(sourceId))
            {
                return $"// Attribution: Unknown source ({sourceId})";
            }

            var source = citations.Sources[sourceId];
            string comment = $"// Attribution: {source.Name}";

            if (!string.IsNullOrEmpty(source.Author))
                comment += $", by {source.Author}";
            if (!string.IsNullOrEmpty(source.Url))
                comment += $"\n// Source: {source.Url}";
            if (!string.IsNullOrEmpty(source.LicenseType))
                comment += $"\n// License: {source.LicenseType}";

            return comment;
        }

        /// <summary>
        /// Generate an attribution header for a file.
        /// </summary>
        /// <param name="filePath">Path to the file</param>
        /// <returns>Formatted header with all attributions for the file</returns>
        public string InsertAttributionHeader(string filePath)
        {
            string relPath = Path.GetRelativePath(projectPath, filePath);
            if (!citations.FileCitations.ContainsKey(relPath))
            {
                return "// No recorded attributions for this file.";
            }

            var fileCitations = citations.FileCitations[relPath];
            var citedSources = new HashSet<string>();
            foreach (var citation in fileCitations)
            {
                foreach (var sourceId in citation.SourceIds)
                {
                    citedSources.Add(sourceId);
                }
            }

            var header = new List<string>
            {
                "/*",
                " * This file contains code derived from or inspired by the following sources:",
                " *"
            };

            foreach (var sourceId in citedSources)
            {
                if (citations.Sources.ContainsKey(sourceId))
                {
                    var source = citations.Sources[sourceId];
                    header.Add($" * - {source.Name}" + (!string.IsNullOrEmpty(source.Url) ? $" ({source.Url})" : ""));
                    if (!string.IsNullOrEmpty(source.Author))
                        header.Add($" *   Author: {source.Author}");
                    if (!string.IsNullOrEmpty(source.LicenseType))
                        header.Add($" *   License: {source.LicenseType}");
                    header.Add(" *");
                }
            }

            header.Add(" */");
            return string.Join("\n", header);
        }

        /// <summary>
        /// Generate a Markdown document with all citations.
        /// </summary>
        /// <returns>Markdown formatted citations document</returns>
        public string GenerateCitationsMarkdown()
        {
            var md = new List<string>
            {
                "# Project Citations",
                "",
                "This document lists all external sources used in this project."
            };

            if (citations.Sources.Count == 0)
            {
                md.Add("\n*No citations recorded.*");
                return string.Join("\n", md);
            }

            md.Add("\n## Sources\n");

            foreach (var source in citations.Sources)
            {
                md.Add($"### {source.Value.Name}");
                if (!string.IsNullOrEmpty(source.Value.Author))
                    md.Add($"**Author:** {source.Value.Author}");
                if (!string.IsNullOrEmpty(source.Value.Url))
                    md.Add($"**URL:** [{source.Value.Url}]({source.Value.Url})");
                if (!string.IsNullOrEmpty(source.Value.LicenseType))
                    md.Add($"**License:** {source.Value.LicenseType}");
                if (!string.IsNullOrEmpty(source.Value.Description))
                    md.Add($"\n{source.Value.Description}");
                md.Add("");
            }

            md.Add("## Usage by File\n");

            foreach (var fileCitation in citations.FileCitations)
            {
                md.Add($"### {fileCitation.Key}");
                foreach (var citation in fileCitation.Value)
                {
                    var sourceNames = new List<string>();
                    foreach (var sourceId in citation.SourceIds)
                    {
                        if (citations.Sources.TryGetValue(sourceId, out var source))
                            sourceNames.Add(source.Name);
                        else
                            sourceNames.Add($"Unknown ({sourceId})");
                    }

                    string locations = "";
                    if (citation.LineStart.HasValue && citation.LineEnd.HasValue)
                        locations = $" (lines {citation.LineStart}-{citation.LineEnd})";
                    else if (citation.LineStart.HasValue)
                        locations = $" (from line {citation.LineStart})";

                    md.Add($"- Uses: {string.Join(", ", sourceNames)}{locations}");
                    if (!string.IsNullOrEmpty(citation.Comment))
                        md.Add($"  - Note: {citation.Comment}");
                }
                md.Add("");
            }

            return string.Join("\n", md);
        }

        /// <summary>
        /// Export citations to a markdown file.
        /// </summary>
        /// <param name="outputPath">Optional path for the markdown file</param>
        /// <returns>Path to the created markdown file</returns>
        public string ExportCitationsMarkdown(string outputPath = null)
        {
            outputPath ??= Path.Combine(projectPath, "CITATIONS.md");
            string mdContent = GenerateCitationsMarkdown();
            File.WriteAllText(outputPath, mdContent);
            return outputPath;
        }

        /// <summary>
        /// Save citations to the citations file.
        /// </summary>
        private void SaveCitations()
        {
            var options = new JsonSerializerOptions
            {
                WriteIndented = true
            };
            string jsonString = JsonSerializer.Serialize(citations, options);
            File.WriteAllText(citationsFile, jsonString);
        }
    }

    #region Data Classes
    public class CitationData
    {
        [JsonPropertyName("project_info")]
        public ProjectInfo ProjectInfo { get; set; }

        [JsonPropertyName("sources")]
        public Dictionary<string, SourceInfo> Sources { get; set; }

        [JsonPropertyName("file_citations")]
        public Dictionary<string, List<Citation>> FileCitations { get; set; }
    }

    public class ProjectInfo
    {
        [JsonPropertyName("name")]
        public string Name { get; set; }

        [JsonPropertyName("created_at")]
        public DateTime CreatedAt { get; set; }
    }

    public class SourceInfo
    {
        [JsonPropertyName("name")]
        public string Name { get; set; }

        [JsonPropertyName("url")]
        public string Url { get; set; }

        [JsonPropertyName("author")]
        public string Author { get; set; }

        [JsonPropertyName("license_type")]
        public string LicenseType { get; set; }

        [JsonPropertyName("description")]
        public string Description { get; set; }

        [JsonPropertyName("added_at")]
        public DateTime AddedAt { get; set; }
    }

    public class Citation
    {
        [JsonPropertyName("source_ids")]
        public string[] SourceIds { get; set; }

        [JsonPropertyName("cited_at")]
        public DateTime CitedAt { get; set; }

        [JsonPropertyName("line_start")]
        public int? LineStart { get; set; }

        [JsonPropertyName("line_end")]
        public int? LineEnd { get; set; }

        [JsonPropertyName("comment")]
        public string Comment { get; set; }
    }
    #endregion
}
